File: android/app/voice/VoiceCommands.kt
package com.gemnav.app.voice

/**
 * Voice command definitions for all tiers
 */
sealed class VoiceCommand {
    // Navigation commands
    data class Navigate(
        val destination: String,
        val waypoints: List<String> = emptyList()
    ) : VoiceCommand()
    
    data class AddStop(val location: String) : VoiceCommand()
    object CancelNavigation : VoiceCommand()
    object RecenterMap : VoiceCommand()
    
    // Audio commands
    object MuteVoice : VoiceCommand()
    object UnmuteVoice : VoiceCommand()
    object RepeatInstruction : VoiceCommand()
    
    // Information queries
    object GetETA : VoiceCommand()
    object GetDistanceRemaining : VoiceCommand()
    object GetTrafficStatus : VoiceCommand()
    data class SearchAlongRoute(
        val query: String,
        val filters: Map<String, String> = emptyMap()
    ) : VoiceCommand()
    
    // Route modifications
    data class AvoidRouteFeature(val feature: RouteFeature) : VoiceCommand()
    object ShowAlternativeRoutes : VoiceCommand()
    data class OptimizeRoute(val criterion: OptimizationCriterion) : VoiceCommand()
    
    // Truck-specific (Pro tier)
    data class FindTruckPOI(val type: TruckPOIType) : VoiceCommand()
    object CheckBridgeClearances : VoiceCommand()
    object CheckHeightRestrictions : VoiceCommand()
    object CheckWeightLimits : VoiceCommand()
    object FindWeighStations : VoiceCommand()
    
    // Conversation
    data class Clarification(val question: String) : VoiceCommand()
    object Unknown : VoiceCommand()
}

/**
 * Route features that can be avoided
 */
enum class RouteFeature {
    TOLLS,
    HIGHWAYS,
    FERRIES,
    UNPAVED_ROADS
}

/**
 * Route optimization criteria
 */
enum class OptimizationCriterion {
    FASTEST,
    SHORTEST,
    ECO_FRIENDLY,
    AVOID_TRAFFIC
}

/**
 * Truck-specific POI types (Pro tier)
 */
enum class TruckPOIType(val displayName: String) {
    TRUCK_STOP("truck stops"),
    WEIGH_STATION("weigh stations"),
    REST_AREA("rest areas with truck parking"),
    DIESEL_STATION("diesel fuel stations"),
    TRUCK_WASH("truck washes"),
    REPAIR_SHOP("truck repair shops")
}

/**
 * Command execution result
 */
sealed class CommandResult {
    data class Success(val message: String, val data: Any? = null) : CommandResult()
    data class Error(val message: String) : CommandResult()
    data class TierRestricted(val message: String) : CommandResult()
    data class NeedsClarification(val question: String) : CommandResult()
}

File: ios/app/voice/VoiceCommands.swift
import Foundation

// MARK: - Voice Commands

enum VoiceCommand {
    // Navigation commands
    case navigate(destination: String, waypoints: [String] = [])
    case addStop(location: String)
    case cancelNavigation
    case recenterMap
    
    // Audio commands
    case muteVoice
    case unmuteVoice
    case repeatInstruction
    
    // Information queries
    case getETA
    case getDistanceRemaining
    case getTrafficStatus
    case searchAlongRoute(query: String, filters: [String: String] = [:])
    
    // Route modifications
    case avoidRouteFeature(feature: RouteFeature)
    case showAlternativeRoutes
    case optimizeRoute(criterion: OptimizationCriterion)
    
    // Truck-specific (Pro tier)
    case findTruckPOI(type: TruckPOIType)
    case checkBridgeClearances
    case checkHeightRestrictions
    case checkWeightLimits
    case findWeighStations
    
    // Conversation
    case clarification(question: String)
    case unknown
}

// MARK: - Route Features

enum RouteFeature: String {
    case tolls
    case highways
    case ferries
    case unpavedRoads
}

// MARK: - Optimization Criteria

enum OptimizationCriterion: String {
    case fastest
    case shortest
    case ecoFriendly
    case avoidTraffic
}

// MARK: - Truck POI Types

enum TruckPOIType: String {
    case truckStop
    case weighStation
    case restArea
    case dieselStation
    case truckWash
    case repairShop
    
    var displayName: String {
        switch self {
        case .truckStop: return "truck stops"
        case .weighStation: return "weigh stations"
        case .restArea: return "rest areas with truck parking"
        case .dieselStation: return "diesel fuel stations"
        case .truckWash: return "truck washes"
        case .repairShop: return "truck repair shops"
        }
    }
}

// MARK: - Command Result

enum CommandResult {
    case success(message: String, data: Any? = nil)
    case error(message: String)
    case tierRestricted(message: String)
    case needsClarification(question: String)
}

// MARK: - Voice State

enum VoiceState: Equatable {
    case idle
    case listening
    case processing
    case speaking(text: String)
    case error(message: String)
    
    static func == (lhs: VoiceState, rhs: VoiceState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle),
             (.listening, .listening),
             (.processing, .processing):
            return true
        case (.speaking(let lhsText), .speaking(let rhsText)):
            return lhsText == rhsText
        case (.error(let lhsMsg), .error(let rhsMsg)):
            return lhsMsg == rhsMsg
        default:
            return false
        }
    }
}

// MARK: - Voice Trigger

enum VoiceTrigger {
    case manual      // User tapped microphone button
    case wakeWord    // Wake word detected
    case continuation // Continuation of multi-turn conversation
}

// MARK: - Conversation Turn

struct ConversationTurn {
    let role: String  // "user" or "assistant"
    let content: String
    let timestamp: Date
    
    init(role: String, content: String, timestamp: Date = Date()) {
        self.role = role
        self.content = content
        self.timestamp = timestamp
    }
}

// MARK: - Subscription Tier Extensions

extension SubscriptionTier {
    var allowsWakeWord: Bool {
        return self == .plus || self == .pro
    }
    
    var allowsAdvancedVoice: Bool {
        return self == .plus || self == .pro
    }
}
